# Artemis: Reads To Results In Less Time
author: "Timothy J. Triche, Jr, Anthony R. Colombo, Harold Pimentel"   
date: "`r Sys.Date()`"   
output: html_document:    
        
          
##Introduction
Kallisto is software developed by Nicolas Bray, Harold Pimentel, Pall Melsted, and Lior Pachter (UC Berkeley) that analyzes 30 million unaligned paired-end RNA-Seq reads in less than 5 minutes on a standard laptop computer.   Kallisto quantifies transcript abundance from input RNA-Seq reads by using a process, known as pseudoalignment, which identifies the read-transcript compatibility matrix. Artemis is a BioConductor package that extends functions and utilities for RNA-Seq analysis from raw reads to results in minutes.

## Loading already-processed transcript abundance estimates

This is also fairly simple:

```R
covs <- DataFrame(sampleDir=c("n1","n2","n4","s1","s2","s4"),
                  normal=c(1, 1, 1, 0, 0, 0),
                  subject=as.factor(c(1, 2, 4, 1, 2, 4)))
rownames(covs) <- covs$sampleDirs
merged <- mergeKallisto(covariates=covs, outputPath="/data/output")
```

The resulting object has annotations for distinct coding and non-coding transcripts,
along with an indication of the general biological category of each transcript bundle
(e.g. protein-coding transcript, pseudogene, ribosomal RNA, etc.).  Transcripts for 
which annotations could not be found are assigned to the chromosome "Unknown" and 
retained as part of the KallistoExperiment, albeit with less informative annotation.


## Gene-wise analysis

One of the most common tasks when analysing the abundance of RNA transcripts is to 
aggregate estimates by "gene" (i.e., "bundle of transcripts with extensive homology")
and group the observed changes by biochemical pathway, gene sets associated with 
experimental outcomes, or perhaps drug metabolism and metabolic targets.  For this,
we typically collapse all the transcripts in each bundle, and run tests across those:

```R
gwa <- geneWiseAnalysis(merged)
```


## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
