# Artemis: Reads To Results In Less Time
author: Timothy J. Triche, Jr, Anthony R. Colombo, Harold Pimentel       
date:  `r Sys.Date()`                 
output: html_document:        
        
          
##Introduction
Kallisto is software developed by Nicolas Bray, Harold Pimentel, Pall Melsted, and Lior Pachter (UC Berkeley) that analyzes 30 million unaligned paired-end RNA-Seq reads in less than 5 minutes on a standard laptop computer.   Kallisto quantifies transcript abundance from input RNA-Seq reads by using a process, known as pseudoalignment, which identifies the read-transcript compatibility matrix. Artemis is a BioConductor package that extends functions and utilities for RNA-Seq analysis from raw reads to results in minutes.

##Reads to Quantification to Annotation
Artemis was designed to reduce the programmative steps required to quantify and annotate multitudes of sample directories. Artemis calls Kallisto to perform on- the-fly transcriptome indexing and quantification recursively for numerous sample directories. For RNA- Seq projects with numerous sequenced samples, Artemis encapsulates expensive preparatory routines. Artemis programmatically orders FASTQ files output from DNA sequencers and inputs a list required by
Kallisto for processing multitudes of demultiplexed reads. The Artemis function 'runKallisto' recursively indexes transcriptomes and quantifies abundances for any number of samples. The function 'mergeKallisto' merges quantified output into an object of ofsubclass a KallistoExperiment-class, SummarizedExperiment-class. Standard mutators and accessor methods from SummarizedExperiment- methods are preserved in KallistoExperiment-methods. Gene annotation is performed from user-selected bundled transcriptomes (ERCC, Ensembl, and/or
RepBase) simultaneously merging annotated samples into one R object: KallistoExperiment. Artemis annotates genes for Homo-Sapiens GrCh38 and Mouse GrCm38 (NCBI). Routines such as 'annotateBundles' yields annotated genes from transcriptomes such as External RNA Control Consortium (ERCC), Ensembl release 81 of non-coding RNA, coding RNA, and a hg38 repeatome for both species.

###Kallisto Installation
For linux systems, after installing the dependencies, kallisto is installed via:
``` mkdir /KallistoSource        
cd /KallistoSource           
git clone https://github.com/pachterlab/kallisto.git      
cd ./kallisto     
mkdir ./build     
cd ./build       
cmake ..              
make          
make install ```       



###Fastq Quantification
Kallisto source must be built locally, and then quantifying the raw fastq's requires calling quant option:

```{r}
library(artemis)      
library(artemisData)       
library(TxDbLite)       
pathBase<-system.file("extdata",package="artemisData")       
FastaPath <- paste0(pathBase, "/fasta")
setwd(FastaPath)
OutputPath<-"/home/anthonycolombo/Git_Repos/artemisData/data/output"
fastqPath <- paste0(pathBase, "/fastq")            
samples <- c(MrN="MrN", MrT="MrT")              
FastaFiles <- c( "ERCC.fa", ## spike-in controls             
                 "Homo_sapiens.RepBase.20_05.humrep.fa", ## repeats         
                 "Homo_sapiens.RepBase.20_05.humsub.fa")  ## ALUs and such             
indexName <- indexKallisto(fastaFiles=FastaFiles, fastaPath=FastaPath)$indexName
library(parallel)
results <- lapply(samples, 
                    runKallisto,
                    indexName=indexName,
                    fastqPath=fastqPath,
                    fastaPath=FastaPath,
                    bootstraps=100,
                    outputPath=OutputPath)
```


## Loading already-processed transcript abundance estimates

This is also fairly simple:

```{r}
merged <- mergeKallisto(samples, outputPath=OutputPath)
design <- DataFrame(sampleDir=c("MrN","MrT"),
                  intercept=c(1,1),
                  normal=c(1, 0))
rownames(design) <- design$sampleDir
```

The resulting object has annotations for distinct coding and non-coding transcripts,
along with an indication of the general biological category of each transcript bundle
(e.g. protein-coding transcript, pseudogene, ribosomal RNA, etc.).  Transcripts for 
which annotations could not be found are assigned to the chromosome "Unknown" and 
retained as part of the KallistoExperiment, albeit with less informative annotation.


## Gene-wise analysis

One of the most common tasks when analysing the abundance of RNA transcripts is to 
aggregate estimates by "gene" (i.e., "bundle of transcripts with extensive homology")
and group the observed changes by biochemical pathway, gene sets associated with 
experimental outcomes, or perhaps drug metabolism and metabolic targets.  For this,
we typically collapse all the transcripts in each bundle, and run tests across those:

```{r}
#gwa <- geneWiseAnalysis(merged,design=design,how="cpm",species="Homo.sapiens")
```


## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
