---
title: "Artemis: Raw Reads To Pathway Analsyes In Much Less Time"
author:  "Timothy J. Triche, Jr, Anthony R. Colombo, Harold Pimentel"
output: 
    pdf_document:
      toc: true
      number_sections: true
date: "`r format(Sys.time(), '%d %B, %Y')`"
---
          
#Introduction
Kallisto is software developed by Nicolas Bray, Harold Pimentel, Pall Melsted, and Lior Pachter (UC Berkeley) that analyzes 30 million unaligned paired-end RNA-Seq reads in less than 5 minutes on a standard laptop computer.   Kallisto quantifies transcript abundance from input RNA-Seq reads by using a process, known as pseudoalignment, which identifies the read-transcript compatibility matrix. Artemis is a BioConductor package that extends functions and utilities for RNA-Seq analysis from raw reads to results in minutes.

#Reads to Quantification to Annotation
Artemis was designed to reduce the programmative steps required to quantify and annotate multitudes of sample directories. Artemis calls Kallisto to perform on- the-fly transcriptome indexing and quantification recursively for numerous sample directories. For RNA- Seq projects with numerous sequenced samples, Artemis encapsulates expensive preparatory routines. Artemis programmatically orders FASTQ files output from DNA sequencers and inputs a list required by
Kallisto for processing multitudes of demultiplexed reads. The Artemis function 'runKallisto' recursively indexes transcriptomes and quantifies abundances for any number of samples. The function 'mergeKallisto' merges quantified output into an object of ofsubclass a KallistoExperiment-class, SummarizedExperiment-class. Standard mutators and accessor methods from SummarizedExperiment- methods are preserved in KallistoExperiment-methods. Gene annotation is performed from user-selected bundled transcriptomes (ERCC, Ensembl, and/or
RepBase) simultaneously merging annotated samples into one R object: KallistoExperiment. Artemis annotates genes for Homo-Sapiens GrCh38 and Mouse GrCm38 (NCBI). Routines such as 'annotateBundles' yields annotated genes from transcriptomes such as External RNA Control Consortium (ERCC), Ensembl release 81 of non-coding RNA, coding RNA, and a hg38 repeatome for both species.

##Kallisto Installation
For linux systems, after installing the dependencies, kallisto is installed via:
```
mkdir /KallistoSource           
cd /KallistoSource             
git clone https://github.com/pachterlab/kallisto.git         
cd ./kallisto       
mkdir ./build        
cd ./build       
cmake ..              
make          
make install   
```     

#Gene Wise Analysis
Artemis supports various levels of analysis, namely transcript-level or gene-level analysis which involves the Limma package for differential expression analysis.   
  Gene Wise Analysis is founded on the idea that groups of transcripts by a fixed Ensembl Gene ID is termed a "gene"; where "gene" counts are defined as the sum of all transcripts identified by the same unique Ensembl Gene Id.  Gene Wise analysis generates bundled and aggregated transcripts associated with a specific Ensembl Gene ID.  Artemis wraps limma around another method titled "collapseBundles", which collapses transcripts into appropriate groups and sums the quantified transcript counts of the group; these transcript aggregated counts are defined as "gene" counts.  

## The Measure Depends On The Level
   Not all transcripts have the same function homology.  Most folks agree that genes are made up by transcripts defined by the transcripts' coordinate location on the genome.  However there are transcipt isoforms in DNMT3A and WT1 that have radically different biological function depending on the transcript isoform that is present.  The problem with conducting *only* a gene level analysis is that many genes can have the same total gene level total quantified counts; however the biological mechanisms for the same "gene" can vary greatly by a single transcript isoform.
 
```{r message=FALSE}
suppressWarnings(suppressPackageStartupMessages(library(artemis)))    
suppressPackageStartupMessages(library(artemisData))       
jsonFile <- system.file("extdata", "NS.JSON", package="artemis")
appSession <- fetchAppSession(jsonFile) ## a
names(appSession$samples) <- appSession$samples ## so column names get set 
appSession$outputPath <- system.file("extdata", "", package="artemisData")
pathBase<-system.file("extdata",package="artemisData")
fastaPath <- paste0(pathBase, "/fasta")
appSession$fastaPath<-fastaPath
NS <- mergeKallisto(appSession$samples, 
                      outputPath=appSession$outputPath)
```

##Creating The Design Matrix
In order to analyze bundle-aggregated transcripts defined as "genes", we create a design matrix which controls for individual effects and contrasts treatment effects across individual subjects.  

```{r}
NS$subject <- factor(substr(colnames(NS), 2, 2))
NS$treatment <- substr(colnames(NS), 1, 1) == "s"
NS$ID <- NULL
design <- with(as(colData(NS), "data.frame"),
                  model.matrix( ~ treatment + subject ))
rownames(design) <- colnames(NS)
metadata(NS)$design <- design
design
```
#Annotate!
In order to run gene-wise analysis, Artemis requires that the merged KallistoExperiment must be annotated; this is because we must collapse transcripts into groups linked to unique Ensembl Gene Ids.    


## Buiding Annotation libraries
  Library Annotations are built using TxDbLite; these annotation databases allow for lite annotations parsing gene names, bio-types and family type from reference fastas from ERCC, Ensembl, or RepBase.  Currently exonic, intronic, or other coordinate dependent information is not included in TxDbLite.   The supplemental package artemisData stores the ready-to-load annotation libraries under /extdata/Libraries directory.  For demonstration, we build the libraries under the artemisData/extdata/fasta/tmp directory.

```{r}
suppressPackageStartupMessages(library(TxDbLite))
suppressWarnings(suppressPackageStartupMessages(library(artemis)))
suppressPackageStartupMessages(library(artemisData))
jsonFile <- system.file("extdata", "NS.JSON", package="artemis")
appSession <- fetchAppSession(jsonFile)
names(appSession$samples) <- appSession$samples
appSession$outputPath <- system.file("extdata", package="artemisData")
fastaPath<-system.file("extdata","fasta",package="artemisData")
appSession$fastaPath<-fastaPath
cd<-appSession$fastaPath
setwd(paste0(appSession$fastaPath,"/","tmp"))
NS <- mergeKallisto(appSession$samples,
                      outputPath=appSession$outputPath)

fastaTx<-c("ERCC.fa.gz","Homo_sapiens.GRCh38.81.cdna.all.fa.gz","Homo_sapiens.RepBase.20_05.merged.fa")
erccDb<-erccDbLiteFromFasta(paste0(appSession$fastaPath,"/tmp/","ERCC.fa.gz"))
erccPkg<-makeErccDbLitePkg(erccDb,destDir=paste0(appSession$fastaPath,"/","tmp"))
#Create a Ensembl Annotation Db with cdna and ncrna
lapply(fastaTx,function(x) findDupes(x))
ensDb<-ensDbLiteFromFasta("Homo_sapiens.GRCh38.81.cdna.all.fa.gz")
ensPkg<-makeEnsDbLitePkg(ensDb,destDir=paste0(appSession$fastaPath,"/","tmp"))
repDb<-repDbLiteFromFasta("Homo_sapiens.RepBase.20_05.merged.fa")
repPkg<-makeRepDbLitePkg(repDb,destDir=paste0(appSession$fastaPath,"/","tmp"))
ErccDbLite(erccDb)
EnsDbLite(ensDb)
RepDbLite(repDb)
transcripts(ErccDbLite(erccDb))
transcripts(EnsDbLite(ensDb))
transcripts(RepDbLite(repDb))

files<-dir(paste0(appSession$fastaPath,"/tmp"))[!dir(paste0(appSession$fastaPath,"/tmp")) %in% fastaTx]

lapply(files,function(x) system(paste0("rm -r ",x)))

```

# Annotating Merged KallistoExperiment Containers
   Artemis has a function "annotateFeatures.R" which annotates ERCC, Ensembl, and RepBase databases for species Homo-Sapiens, Mus-musculus, and Rattus norvegicus.  The method "annotateFeatures.R" annotates the merged KallistoExperiment against every TxDbLite library simulatenously. These annotation databases are defined as 'lite' because they do not store exonic or intronic coordinates.
```{r}
suppressPackageStartupMessages(library(artemis))
library(artemisData)
suppressPackageStartupMessages(library(TxDbLite))
samples<-c("n1","n2","n4","s1","s2","s4")
pathBase<-system.file("extdata",package="artemisData")
merged <- mergeKallisto(samples, outputPath=pathBase)
libraryPath<-system.file("extdata","Libraries",package="artemisData")
command<-paste0("sudo R CMD INSTALL ",libraryPath,"/",dir(libraryPath))
lapply(command,function(x) system(x))
merged<-annotateFeatures(merged, level="transcript") #annotate features using transcriptomes
NS<-suppressWarnings(annotateFeatures(NS,level="transcript"))
NS$subject <- factor(substr(colnames(NS), 2, 2))
NS$treatment <- substr(colnames(NS), 1, 1) == "s"
NS$ID <- NULL
design <- with(as(colData(NS), "data.frame"),
                  model.matrix( ~ treatment + subject ))
rownames(design) <- colnames(NS)
metadata(NS)$design <- design
#returns a KallistoExperiment at the gene level
GWA<-geneWiseAnalysis(NS,design=design,
                       how="cpm",
                       p.cutoff=0.05,
                       fold.cutoff=1,
                       read.cutoff=1,
                       species="Homo.sapiens")

head(GWA$limmaWithMeta,n=20)
```
#Gene Wise Analysis
 Gene wise analysis collapses transcripts into groups related to specific ensembl "gene" Ids.  The package TxDbLite parses the Ensembl, or RepBase transcript fasta files and stores the respective gene id's associated with the given transcript documented in the transcript fasta header.  Artemis' method for gene wise analysis calls "collapseBundles.R" which then calculates the aggregated total counts of transcripts for each unique gene id association.  Thus the "gene" count is defined as the sum of all quantified transcripts associated with a specific gene identifier.  



##Understanding Gene Wise Analysis Output
  The output contains a list of limma derived expression values, and enrichment data derived by biomaRt.
##Expression Results    
  The expression results were generated by limma/voom and have the meta biotype, gene name, etc information included in the gene wise analysis results.

##Understanding Gene Wise Analysis Output
  The output contains a list of limma derived expression values, and entrezID, gene name, and gene biotypes derived by biomaRt and TxDbLite respectively.  The expression results were generated by limma/voom and have the meta biotype, gene name, etc information included in the gene wise analysis results.
