f--
title: "Artemis: reads to results in a lot less time"
author: "Timothy J. Triche, Jr."
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Loading already-processed transcript abundance estimates

This is also fairly simple:

```R
covs <- DataFrame(sampleDir=c("n1","n2","n4","s1","s2","s4"),
                  normal=c(1, 1, 1, 0, 0, 0),
                  subject=as.factor(c(1, 2, 4, 1, 2, 4)))
rownames(covs) <- covs$sampleDirs
merged <- mergeKallisto(covariates=covs, outputPath="/data/output")
```

The resulting object has annotations for distinct coding and non-coding transcripts,
along with an indication of the general biological category of each transcript bundle
(e.g. protein-coding transcript, pseudogene, ribosomal RNA, etc.).  Transcripts for 
which annotations could not be found are assigned to the chromosome "Unknown" and 
retained as part of the KallistoExperiment, albeit with less informative annotation.


## Gene-wise analysis

One of the most common tasks when analysing the abundance of RNA transcripts is to 
aggregate estimates by "gene" (i.e., "bundle of transcripts with extensive homology")
and group the observed changes by biochemical pathway, gene sets associated with 
experimental outcomes, or perhaps drug metabolism and metabolic targets.  For this,
we typically collapse all the transcripts in each bundle, and run tests across those:

```R
gwa <- geneWiseAnalysis(merged)
```


## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
